chain/cosmos/chain_node.go

```diff
@@ -285,29 +298,65 @@ func (tn *ChainNode) FindTxs(ctx context.Context, height uint64) ([]blockdb.Tx,
 			tn.logger().Info("Failed to marshal tx to json", zap.Uint64("height", height), zap.Error(err))
 			continue
 		}
-		txs[i].Data = b
+		newTx.Data = b
 
-		// Request the transaction directly in order to get the tendermint events.
-		txRes, err := tn.Client.Tx(ctx, tx.Hash(), false)
-		if err != nil {
-			tn.logger().Info("Failed to retrieve tx", zap.Uint64("height", height), zap.Error(err))
-			continue
-		}
+		rTx := blockRes.TxsResults[i]
 
-		txs[i].Events = make([]blockdb.Event, len(txRes.TxResult.Events))
-		for j, e := range txRes.TxResult.Events {
+		newTx.Events = make([]blockdb.Event, len(rTx.Events))
+		for j, e := range rTx.Events {
 			attrs := make([]blockdb.EventAttribute, len(e.Attributes))
 			for k, attr := range e.Attributes {
 				attrs[k] = blockdb.EventAttribute{
 					Key:   string(attr.Key),
 					Value: string(attr.Value),
 				}
 			}
-			txs[i].Events[j] = blockdb.Event{
+			newTx.Events[j] = blockdb.Event{
+				Type:       e.Type,
+				Attributes: attrs,
+			}
+		}
+		txs = append(txs, newTx)
+	}
+	if len(blockRes.BeginBlockEvents) > 0 {
+		beginBlockTx := blockdb.Tx{
+			Data: []byte(`{"data":"begin_block"}`),
```

Comment:
So we are artificially creating a tx in order to query these events?

Would this confuse a newcomer to cosmos who would then expect to see these txs on the blockchain? (i.e. if they look in mintscan). 

I didn't realize there were events not associated with a tx in cosmos. 

I'm ok with this as a workaround/hack, but perhaps the long-term solution is saving it in `tendermint_event` table which may or may not have a relationship to a tx but always has a relationship to a block. IIUC, these begin and end block events are 0:many to a block.

Reply:
You could also extend the `{"data": <begin/end block>, "note": "this is a transaction artificially created for debugging purposes"}`

Reply:
I considered schema changes, but since it's really only for debugging, the effort did not seem necessary. Yes, the events are 0:many just like a typical tx, so it seemed to plug in well to the existing queries. Good call about the `note` property to aid in understanding, added.